### PR TITLE
Return empty dataframe when reading a Parquet file using empty `columns` option

### LINKED ISSUE #8668: [BUG] Empty `columns` is not returning empty dataframe in `read_parquet`
**Describe the bug**
`columns` is used to select what columns to be read from parquet file. When `columns` is `None`, all columns are read. When `columns` is `[]`, no columns should be read, but in `cudf` all `[]` and `None` have the same behavior.

**Steps/Code to reproduce bug**
```python
>>> import cudf
>>> df = cudf.DataFrame({'a':[1, 2, 3], 'b':[4, 5, 6]})
>>> df.to_parquet('a')
>>> import pandas as pd
>>> pd.read_parquet('a')
   a  b
0  1  4
1  2  5
2  3  6
>>> pd.read_parquet('a', columns=None)
   a  b
0  1  4
1  2  5
2  3  6
>>> pd.read_parquet('a', columns=[])
Empty DataFrame
Columns: []
Index: []
>>> cudf.read_parquet('a', columns=None)
   a  b
0  1  4
1  2  5
2  3  6
>>> cudf.read_parquet('a', columns=[])
   a  b
0  1  4
1  2  5
2  3  6
```
**Expected behavior**

**Environment overview (please complete the following information)**
 - Environment location: Bare-metal
 - Method of cuDF install: from source

### PR DESCRIPTION
**Pull request by vuule**
Fixes #8668
Store the columns option as `optional`:

- `nullopt` when columns are not passed by caller - read all columns.
- Empty vector when caller explicitly passes an empty list/vector - return empty dataframe.
- Vector of column names - read columns with given names.

### CODE REVIEW

File: python/cudf/cudf/_lib/parquet.pyx [hunk 1]
```diff
@@ -165,6 +159,12 @@ cpdef read_parquet(filepaths_or_buffers, columns=None, row_groups=None,
         .timestamp_type(cpp_timestamp_type)
         .build()
     )
+    cdef vector[string] cpp_columns
```
Comment (galipremsagar) [inline python/cudf/cudf/_lib/parquet.pyx:162]:
```suggestion
    cdef vector[string] cpp_columns
    allow_range_index = True
```

File: python/cudf/cudf/_lib/parquet.pyx [hunk 2]
```diff
@@ -165,6 +159,12 @@ cpdef read_parquet(filepaths_or_buffers, columns=None, row_groups=None,
...
+    if columns is not None:
+        cpp_columns.reserve(len(columns))
+        for col in columns or []:
+            cpp_columns.push_back(str(col).encode())
```
Comment (galipremsagar) [inline python/cudf/cudf/_lib/parquet.pyx:166]:
```suggestion
        cpp_columns.reserve(len(columns))
        if len(cpp_columns) == 0:
            allow_range_index = False
        for col in columns or []:
            cpp_columns.push_back(str(col).encode())
```

File: python/cudf/cudf/_lib/parquet.pyx [hunk 3]
```diff
@@ -165,6 +159,12 @@ cpdef read_parquet(filepaths_or_buffers, columns=None, row_groups=None,
...
+        args.set_columns(cpp_columns)
```
Comment (galipremsagar) [inline python/cudf/cudf/_lib/parquet.pyx:168]: and then after line 220, inside `if is_range_index:` block, you need to add this condition:
```python
if not allow_index:
    return df
```

It would look like:
```python
# Set the index column
if index_col is not None and len(index_col) > 0:
    if is_range_index:
        if not allow_index:
            return df
        range_index_meta = index_col[0]
```

File: python/cudf/cudf/_lib/parquet.pyx [hunk 4]
```diff
@@ -165,6 +159,12 @@ cpdef read_parquet(filepaths_or_buffers, columns=None, row_groups=None,
...
 
     # Read Parquet
     cdef cudf_io_types.table_with_metadata c_out_table
```
Comment (galipremsagar) [inline python/cudf/cudf/_lib/parquet.pyx:170]: I've also written out an extension to the pytest you sent me offline. Here is how we can validate all the conditions:
```python
@pytest.mark.parametrize("index", [True, False, None])
@pytest.mark.parametrize("columns", [None, [], ["b", "a"]])
def test_parquet_columns_and_index_param(index, columns):
    buffer = BytesIO()
    df = cudf.DataFrame({"a": [1, 2, 3], "b": ["a", "b", "c"]})
    df.to_parquet(buffer, index=index)

    expected = pd.read_parquet(buffer, columns=columns)
    got = cudf.read_parquet(buffer, columns=columns)

    assert_eq(expected, got, check_index_type=True)
```

**Review by galipremsagar — CHANGES_REQUESTED** [review]
I've tried out your changes locally, `libcudf` behavior seems to be expected now.
These are the corresponding additional cython/python changes needed to achieve feature parity with pandas.

**Review by PointKernel — APPROVED** [review]

**Review by devavret — APPROVED** [review]

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 5]
```diff
@@ -956,42 +956,30 @@ public static Table readJSON(Schema schema, JSONOptions opts, HostMemoryBuffer b
...
-  }
-
   /**
    * Read parquet formatted data.
    * @param opts various parquet parsing options.
    * @param buffer raw parquet formatted bytes.
    * @return the data parsed as a table on the GPU.
    */
   public static Table readParquet(ParquetOptions opts, byte[] buffer) {
+    if (opts.getIncludeColumnNames().length == 0) {
```
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/Table.java:980]: This looks like an undesirable change and is reflected in the corresponding update to the tests.  If the caller must specify all columns to read from the file, how does the caller read a Parquet file when they don't know the columns in the file?  (e.g.: doing dynamic schema discovery). This used to be possible with the Java API but now seems like it isn't.

**Review by jlowe — CHANGES_REQUESTED** [review]

File: java/src/test/java/ai/rapids/cudf/TableTest.java [hunk 6]
```diff
@@ -588,7 +588,35 @@ void testReadParquetBuffer() throws IOException {
...
+            .includeColumn("occupancy_status")
+            .includeColumn("property_state")
+            .includeColumn("zip")
+            .includeColumn("mortgage_insurance_percent")
+            .includeColumn("product_type")
+            .includeColumn("coborrow_credit_score")
+            .includeColumn("mortgage_insurance_type")
+            .includeColumn("relocation_mortgage_indicator")
+            .includeColumn("quarter")
+            .includeColumn("seller_id")
```
[the remark this answers is not in the document]
Comment (vuule) [reply java/src/test/java/ai/rapids/cudf/TableTest.java:617]: I'm not sure how to achieve this using `includeColumn`, but having `includeColumns` that takes an array of names allows this behavior - skip the call to `includeColumns` if you want to read all columns. If you want to read none, pass an empty array.

**Comment by andygrove** [comment]
I misunderstood what was needed on the Java side here. I have re-implemented this in my branch at https://github.com/andygrove/cudf/tree/bug-pq-reader-empty-col-names-java2 so that the JNI code will not add columns to the `parquet_reader_options::builder` if there are no columns.

@vuule could you merge those changes into this PR?

File: java/src/test/java/ai/rapids/cudf/TableTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jlowe) [reply java/src/test/java/ai/rapids/cudf/TableTest.java:617]: Ideally this change should not have a significant impact on the API.  If no columns are specified (i.e.:  no `includeColumn`  or `includeColumns` calls) then it should read all columns as it did before.  That means all the existing tests should run without modification.  However the code explicitly throws if no columns were ever specified, which required many existing tests to be updated.  That's the thing we need to fix, otherwise there's no way to load a Parquet file without knowing the columns up-front.

> having includeColumns that takes an array of names allows this behavior - skip the call to includeColumns if you want to read all columns. If you want to read none, pass an empty array.

Or don't call `includeColumn` or `includeColumns` at all, which is the way it used to work.  If we want to add special behavior for a case where the caller _explicitly_ does not want to load any columns (e.g.: `opts.includeColumns(new String[0])` then we can do so.  However I'm not sure we can actually load an empty `Table`.  `Table` wraps a `cudf::table_view` and the latter does not allow construction with zero columns.  It seems like the Java change, if any, should be checking for no columns specified when `includeColumns` is called and throwing there rather than checking for no columns when trying to load the Parquet file, as the latter means we want to load all columns rather than no columns.  If that's done then no existing tests would be modified, all we would do is add a new test that verifies the API throws in that corner case.

**Review by jlowe — APPROVED** [review]
Assuming the Java tests pass, lgtm.

File: java/src/main/java/ai/rapids/cudf/ColumnFilterOptions.java [hunk 7]
```diff
@@ -21,7 +21,6 @@
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
-import java.util.function.Supplier;
```
Comment (andygrove) [inline java/src/main/java/ai/rapids/cudf/ColumnFilterOptions.java:24]: This was unrelated to this PR and removes an unused import. We can revert this change if needed.

Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ColumnFilterOptions.java:24]: Nit: needs updated copyrights if we're going to modify this file.

Comment (vuule) [reply java/src/main/java/ai/rapids/cudf/ColumnFilterOptions.java:24]: updated

**Comment by vuule** [comment]
@gpucibot merge
